src/dfx/src/lib/extension/manager/mod.rs

```diff
@@ -0,0 +1,86 @@
+use crate::config::cache::{get_bin_cache, get_cache_root};
+use crate::lib::environment::Environment;
+use crate::lib::error::{CacheError, DfxError, DfxResult, ExtensionError};
+use crate::lib::extension::manifest::MANIFEST_FILE_NAME;
+
+use semver::Version;
+use std::path::PathBuf;
+
+use super::manifest::ExtensionManifest;
+use super::Extension;
+
+mod execute;
+mod install;
+mod list;
+mod uninstall;
+
+pub struct ExtensionManager {
+    pub dir: PathBuf,
+    pub dfx_version: Version,
+}
+
+impl ExtensionManager {
+    pub fn new(env: &dyn Environment) -> DfxResult<Self> {
+        let Ok(x) = get_bin_cache(env.get_version().to_string().as_str())
+         else {
+            return Err(DfxError::new(CacheError::FindCacheDirectoryFailed(
+                get_cache_root()?.join("versions"),
+            )))
+        };
+        let dir = x.join("extensions");
+        if !dir.exists() {
+            if let Err(_e) = std::fs::create_dir_all(&dir) {
+                return Err(DfxError::new(
+                    ExtensionError::CreateExtensionDirectoryFailed(dir),
+                ));
+            }
+        }
+        if !dir.is_dir() {
+            return Err(DfxError::new(
+                ExtensionError::ExtensionsDirectoryIsNotADirectory,
+            ));
+        }
```

Comment:
The `dfx_core::fs::ensure_dir_exists` helper can shorten this, as well as include path information in case of an error:
```suggestion
        dfx_core::fs::ensure_dir_exists(&dir)
            .map_err(ExtensionError::EnsureExtensionDirExists)?;
        // error type: EnsureExtensionDirExists(IoError)
```